Allow dynamic shapes of `tuple` type for inputs of `dataclass` type

BoyuanFeng opened the pull request:
Summary:
In `torch.export.export(f, args, kwargs, ..., dynamic_shpapes=None, ...)`, `dataclass` is an acceptable type of inputs (for args and kwargs). The `dynamic_shapes` of the `dataclass` inputs needs to be the same `dataclass` type which replaces each tensor attributes with `dynamic_shapes` of the corresponding tensors. (https://github.com/pytorch/pytorch/blob/main/torch/export/dynamic_shapes.py#L375)

However, some `dataclass` may have limitations on the types of attributes (e.g., having to be tensors) such that the same `dataclass` cannot be constructed for dynamic shapes.

For an input of `dataclass` type, this task enables a `dynamic_shapes` of a tuple type that specifies dynamic shape specifications for each tensor of the input in the same order as the input dataclass type's flatten_fn (https://github.com/pytorch/pytorch/blob/main/torch/utils/_pytree.py#L103)

Differential Revision: D52787912

avikchaudhuri commented on torch/export/dynamic_shapes.py:
```diff
@@ -371,6 +372,11 @@ def tree_zip(combined_args, dynamic_shapes):
                 )
             for k, shape in dynamic_shapes.items():
                 yield from tree_zip(combined_args[k], shape)
+        elif dataclasses.is_dataclass(combined_args) and isinstance(dynamic_shapes, (tuple, list)):
```
I would prefer that we do this not just for dataclasses but all classes that are registered with pytree. In other words, check that `type(combined_args)` is in `SUPPORTED_NODES` and assert that `dynamic_shapes` is `list`, with an error message explaining that for user-registered classes we require you to match the flattened structure. 

Then cut the case below (that checks whether `type(dynamic_shapes) == type(combined_args)`) because that will not be possible in general and is therefore confusing.

Finally, add a test for a regular class (or better, just use KJT in the test).

BoyuanFeng committed:
Allow dynamic shapes of `tuple` type for inputs of `dataclass` type (#117567)

Summary:

In `torch.export.export(f, args, kwargs, ..., dynamic_shpapes=None, ...)`, `dataclass` is an acceptable type of inputs (for args and kwargs). The `dynamic_shapes` of the `dataclass` inputs needs to be the same `dataclass` type which replaces each tensor attributes with `dynamic_shapes` of the corresponding tensors. (https://github.com/pytorch/pytorch/blob/main/torch/export/dynamic_shapes.py#L375)

However, some `dataclass` may have limitations on the types of attributes (e.g., having to be tensors) such that the same `dataclass` cannot be constructed for dynamic shapes.

For an input of `dataclass` type, this task enables a `dynamic_shapes` of a tuple type that specifies dynamic shape specifications for each tensor of the input in the same order as the input dataclass type's flatten_fn (https://github.com/pytorch/pytorch/blob/main/torch/utils/_pytree.py#L103)

Test Plan: buck test //caffe2/test:test_export

Reviewed By: avikchaudhuri

Differential Revision: D52787912
